aidial_sdk/application.py

```diff
@@ -201,6 +219,14 @@ async def _handler(original_request: Request):
             stream = response._generate_stream(impl.chat_completion)
 
             if request.stream:
+                if heartbeat_timeout:
+                    stream = add_heartbeat(
+                        stream,
+                        timeout=heartbeat_timeout,
+                        heartbeat_callback=lambda: log_debug("heartbeat"),
+                        heartbeat_object=": heartbeat\n\n",
```

Comment:
Have you considered to use `event` field, so we could distinguish this in event-stream, like:

```
event: heartbeat
data:
```

Reply:
The problem is that `openai` library isn't aware of the semantics of this "heartbeat" event, which is our invention. 

`openai` library collects **every single** `data:` entry and assumes it to be a stream of valid ChatCompletion JSON chunks. It expects the stream to be data-only: https://github.com/openai/openai-python/blob/891e1c17b7fecbae34d1915ba90c15ddece807f9/src/openai/types/completion_create_params.py#L179

And if the events do occure - it injects it into the stream as `{"event": event, "data": data}` dictionaries, which then are visible to the user: https://github.com/openai/openai-python/blob/891e1c17b7fecbae34d1915ba90c15ddece807f9/src/openai/_streaming.py#L199

But if the user isn't aware of these exotic chunks, he will get runtime errors trying to access non-existing fields like `choices`.

However, openai does respect the comments: https://github.com/openai/openai-python/blob/main/src/openai/_streaming.py#L344

And the comments are the recommended way to support such functionality anyway: https://developer.mozilla.org/en-US/docs/Web/API/Server-sent_events/Using_server-sent_events#event_stream_format